frontend/src/metabase/static-viz/components/Funnel/Funnel.tsx

```diff
@@ -0,0 +1,180 @@
+import React from "react";
+import { Line, Polygon } from "@visx/shape";
+import { Group } from "@visx/group";
+import { Text } from "metabase/static-viz/components/Text";
+import {
+  measureText,
+  measureTextHeight,
+  truncateText,
+} from "metabase/static-viz/lib/text";
+import {
+  FunnelDatum,
+  FunnelSettings,
+} from "metabase/static-viz/components/Funnel/types";
+import {
+  calculateFunnelPolygonPoints,
+  calculateFunnelSteps,
+  calculateStepOpacity,
+  formatPercent,
+} from "metabase/static-viz/components/Funnel/utils";
+
+const layout = {
+  width: 540,
+  height: 300,
+  stepFontSize: 14,
+  measureFontSize: 11,
+  percentFontSize: 16,
+  initialMeasureFontSize: 24,
+  nameFontSize: 16,
+  stepTextOffset: 8,
+  colors: {
+    textMedium: "#949aab",
+    brand: "#509ee3",
+    border: "#f0f0f0",
+  },
+  paddingLeft: 10,
+  stepTopOffset: 10,
+  measureBottomOffset: 8,
+  percentBottomOffset: 24,
+};
+
+const firstMeasureCharWidth = 15;
+const percentCharWidth = 7;
+const stepNameCharWidth = 8;
+
+type FunnelProps = {
+  data: FunnelDatum[];
+  settings: FunnelSettings;
+};
+
+const Funnel = ({ data, settings }: FunnelProps) => {
+  const palette = { ...layout.colors, ...settings.colors };
```

Comment:
Just a though - since we aren't able to add meaningful component tests with jest, would it make sense to extract as much code from components as possible to util functions?